Added: Add support for using image as background

Kruna1Pate1 opened the pull request:
This changes allows the user to use an image as the terminal's background.


## Background Image
To set an image as a background, open the context menu by long clicking on the terminal and select **Menu -> Style -> Set background image.**

To remove background image **Menu -> Style -> Remove background image** (It will also delete generated image files).

When an image is selected from the gallery, it is resized corresponding to device resolution and then compressed into a jpeg. Two image files, `background.jpeg` and `background_landscape.jpeg` are generated and saved in `.termux` folder. Both files are used for different orientations.

> **Note**: Resized images are **center-cropped** generated using [ThumbnailUtils](https://developer.android.com/reference/android/media/ThumbnailUtils#extractThumbnail(android.graphics.Bitmap,%20int,%20int))


## Overlay Color

The overlay color is loaded from the `termux.properties` file with the key `background-overlay-color`. The default value is set to `#59000000`. To change the overlay color, add `background-overlay-color=<color>` to the `termux.properties` file. Supported color formats are `#AARRGGBB` and `#RRGGBB`. If the color is in #RRGGBB format then the alpha value of default color **0x59 (35%)** is used.


## Termux Settings

Background image loading can also be enabled/disabled from settings. From **Termux Settings -> Termux -> Termux Style -> Background Image.** If some malformed image is stored as background image and not letting open termux, then setting can be used to disable image loading.

> **Note**: Disabling the image from the setting will not delete image files and can be restored. 

Setting can also be used to temporarily disable image loading. You can enable it until and unless you delete image files from `.termux` folder.


## Backup

Image files are stored in `.termux` folder, and overlay color value is stored in `termux.properties` and both are part of the backup, so you can easily **Backup** them.

However, the preference for background, whether it's image or color is stored and read from `com.termux_preference.xml` which is not part of the backup. The default value set for the `background_image_enabled` is `false`.

To overcome this issue, if both `background.jpeg` and `background_landscape.jpeg` files are present with the required resolution, then the user can restore them. When again selecting an image from the gallery using `Set background image`, the alert is shown to restore images.  It can also be done from the setting by enabling background image loading.

kernelcornlts commented:
Some feedbacks regarding this change, so far the performance is very good and creates some terminal aesthetics. so far so good!
**Landscape:**
![Screenshot_2022-11-05-01-16-45-48_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/113356647/200036702-4fd35692-0f1a-44c1-ad8e-0af3b1b65ea4.jpg)

**Portrait:**
![Screenshot_2022-11-05-01-16-33-26_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/113356647/200036718-61c04920-4266-4600-aeca-525db8036bde.jpg)

Damaj301damaj-lol commented:
> Some feedbacks regarding this change, so far the performance is very good and creates some terminal aesthetics. so far so good!
> **Landscape:**
> ![Screenshot_2022-11-05-01-16-45-48_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/113356647/200036702-4fd35692-0f1a-44c1-ad8e-0af3b1b65ea4.jpg)
> 
> **Portrait:**
> ![Screenshot_2022-11-05-01-16-33-26_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/113356647/200036718-61c04920-4266-4600-aeca-525db8036bde.jpg)
> 

It would be cool if the bottom keys were also changed

ravener commented:
Could the notch (aka the display cutout) also be filled in landscape?

Kruna1Pate1 commented:
Working on toolbar transparency. Will commit once the proper code is changed.
![Screenshot_2022-11-04-23-47-45-51_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/33123061/200048206-4e4cf625-6b2c-47d0-be3e-15aabe70f90d.jpg)

Kruna1Pate1 commented:
Working on toolbar transparency. will commit once the proper changes are made.


> Could the notch (aka the display cutout) also be filled in landscape?

No, currently it's not filling cutout on landscape orientation. Will see what can be done.

kernelcornlts commented:
I think leaving the notch as is would be better, considering not all devices have notch and the transparency would be enough to fill the gap to make the background consistent.

ravener commented:
> I think leaving the notch as is would be better, considering not all devices have notch and the transparency would be enough to fill the gap to make the background consistent in the terminal.

Yeah nevermind, Termux is not fully full screen so that looks a bit odd

Susko3 commented on termux-shared/src/main/java/com/termux/shared/image/ImageUtils.java:
```diff
@@ -0,0 +1,252 @@
+    public static final String ANY_IMAGE_TYPE = IMAGE_TYPE + "/*";
+
+
+    private static final String LOG_TAG = "FileUtils";
```
```suggestion
    private static final String LOG_TAG = "ImageUtils";
```

kernelcornlts commented:
The extra keys toolbar transparency seems to work like a charm!
![Screenshot_2022-11-05-19-40-19-88_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/113356647/200118169-3e3dc8d0-ac22-4465-bdee-763b16cfee74.jpg)

kyzsuukii commented:
I tried this feature on my device and it doesn't work
the background doesn't change and looks messed up on the text

![Screenshot_2022-11-07-13-28-01-986_com termux](https://user-images.githubusercontent.com/110718021/200233894-c4ddfb25-6b0f-4cbc-9688-d007669275a0.jpg)

![Screenshot_2022-11-07-13-28-43-778_com termux](https://user-images.githubusercontent.com/110718021/200233920-122eac0e-8b19-4b8b-afca-76cce8586c3f.jpg)

Kruna1Pate1 commented:
> I tried this feature on my device and it doesn't work the background doesn't change and looks messed up on the text

can you check if background files are generated properly at `.termux/background.jpeg` and `.termux/background_landscape.jpeg` 

What is the status in setting, is the switch enabled at `Termux Settings -> Termux Style -> Background Image enabled`

kyzsuukii commented:
> > I tried this feature on my device and it doesn't work the background doesn't change and looks messed up on the text
> 
> can you check if background files are generated properly at `.termux/background.jpeg` and `.termux/background_landscape.jpeg`
> 
> What is the status in setting, is the switch enabled at `Termux Settings -> Termux Style -> Background Image enabled`

perfectly available

![Screenshot_2022-11-07-14-11-48-706_bin mt plus canary](https://user-images.githubusercontent.com/110718021/200238306-e64282da-88d2-4855-9e80-69d11d180723.jpg)

kyzsuukii commented:
i try again from action. i think its fixed
but still have same issue

![Screenshot_2022-11-07-16-57-28-114_bin mt plus canary](https://user-images.githubusercontent.com/110718021/200269157-d223b3b8-31c4-4872-9d0e-c5e31db37cfc.jpg)
![Screenshot_2022-11-07-16-56-03-627_com termux](https://user-images.githubusercontent.com/110718021/200269183-f2ebd9d3-1669-4d33-be2d-342b63ad3228.jpg)
![Screenshot_2022-11-07-16-55-40-383_com termux](https://user-images.githubusercontent.com/110718021/200269263-c0957325-d67d-4c05-96b9-112122c098e0.jpg)

Kruna1Pate1 commented:
> i try again from action. i think its fixed but still have same issue

Fixed the issue,
`PorterDuff.Mode.MULTIPLY` was causing the issue. For Android > 9 `BlendMode` filter is used, that's why it was working fine for android > 9.
Now changed to `PorterDuff.Mode.DARKEN` which seems to work pretty fine.

johannesmolla commented:
It amazing feature thank you man

rihaq commented:
@Kruna1Pate1 friend Transparent theme like Windows Terminal when ?
![windows-terminal-acrylic-transparent-background](https://user-images.githubusercontent.com/93857023/201646276-45417a7b-b25c-4f99-94cb-880f283a2023.jpg)
 ?

rihaq commented:
Please add TIOCGWINSZ https://github.com/termux/termux-app/issues/3002

Kruna1Pate1 commented:
> @Kruna1Pate1 friend Transparent theme like Windows Terminal when ?

@rihaq Actually, I have implemented it. But, there are several issues with it. To make an Activity transparent, Activity needs to be `Translucent`.

But, it leads to `android:screenOrientation` set to `unspecified`. Which means it chooses orientation from background activity. And most launchers are fixed to portrait.

Another problem is, the method to change Activity from `opaque` to `translucent` and vice-versa was added in API 30 ["setTranslucent(boolean)"](https://developer.android.com/reference/android/app/Activity#setTranslucent(boolean)). For API <= 29, I can't find a proper way to change it programmatically.

It might create problems for Android 8 devices, because of this [Prevent non-fullscreen activities from influencing orientation](https://android.googlesource.com/platform/frameworks/base/+/android-8.0.0_r1/services/core/java/com/android/server/am/ActivityRecord.java#2196) which was fixed in [Android 8.1](https://android.googlesource.com/platform/frameworks/base/+/android-8.1.0_r1/services/core/java/com/android/server/am/ActivityRecord.java#2193).

Will see if this can be figured out. Also, it looks a little bit uncomfortable on Android. Maybe playing with some alpha value and blur effect can help.


![ss](https://user-images.githubusercontent.com/33123061/201832715-15754aeb-5882-4643-91a0-c4a69fce1386.png)

<br />

> Please add TIOCGWINSZ #3002

I am not much familiar with Terminal code, But will see if I can help there.

HardcodedCat commented:
Here's a suggestion: You could implement background blur that works on Android <11 with [RealtimeBlurView lib](https://github.com/HardcodedCat/RealtimeBlurView_v2). (suggesting mine because official one is very outdated and doesn't build anymore)

But of course, there would be some performance impact, so this could be an optional setting

Kruna1Pate1 commented:
> Here's a suggestion: You could implement background blur along with transparent background that works on Android <11 with [RealtimeBlurView lib](https://github.com/HardcodedCat/RealtimeBlurView_v2). (suggesting mine because official one is very outdated and doesn't build anymore)

I tried it (on Android 12 btw). Looks pretty cool with the blur effect. But as I said have to make the activity `translucent` which causes some trouble. (Can't make it optional because changing it dynamically is added in API 30 as I said).

<br />

> But of course, there would be some performance impact, so this could be an optional setting

Yes, I have noticed that causing some performance drawbacks.

rihaq commented:
Make transparent theme disabled by default, so whose phone is android >11 they can enable it

tareksander commented:
> But as I said have to make the activity `translucent` which causes some trouble. (Can't make it optional because changing it dynamically is added in API 30 as I said).

What I'd do for these cases is have 2 Activities that only differ in the translucency, and make the launcher icon start an invisible helper Activity that starts the translucent or opaque Activity depending on the setting. When you make the original Activity a base class it should be possible without much code duplication. I don't know if Termux can manage having 2 terminal activities, and changing the setting would have to finish the old Activity, so there aren't 2 at the same time.

Kruna1Pate1 commented:
I will try to do that. and i think new pr would be better for this since many changes would require.

Kruna1Pate1 commented:
Make sure to download from [GitHub Action build](https://github.com/termux/termux-app/actions/runs/3409602990).

Damaj301damaj-lol commented:
![Screenshot_20221211-165436_One UI Home.png](https://user-images.githubusercontent.com/90093996/206910865-f7493ff3-5e2f-4ec9-9216-a95968b47e52.png)

rihaq commented:
> > @Kruna1Pate1 friend Transparent theme like Windows Terminal when ?
> > ![windows-terminal-acrylic-transparent-background](https://user-images.githubusercontent.com/93857023/201646276-45417a7b-b25c-4f99-94cb-880f283a2023.jpg)
> > ?
> 
> Use Termux:Float for this

Says the one using Vanced 😂 (I also have it installed for emergency purpose but I use revanced)

No I won't use float shit

Damaj301damaj-lol commented:
> > > @Kruna1Pate1 friend Transparent theme like Windows Terminal when ?
> > > ![windows-terminal-acrylic-transparent-background](https://user-images.githubusercontent.com/93857023/201646276-45417a7b-b25c-4f99-94cb-880f283a2023.jpg)
> > > ?
> > 
> > Use Termux:Float for this
> 
> Says the one using Vanced 😂 (I also have it installed for emergency purpose but I use revanced)
> 
> No I won't use float shit

~then write it yourself~

rihaq commented:
> > > > @Kruna1Pate1 friend Transparent theme like Windows Terminal when ?
> > > > ![windows-terminal-acrylic-transparent-background](https://user-images.githubusercontent.com/93857023/201646276-45417a7b-b25c-4f99-94cb-880f283a2023.jpg)
> > > > ?
> > > 
> > > 
> > > Use Termux:Float for this
> > 
> > 
> > Says the one using Vanced 😂 (I also have it installed for emergency purpose but I use revanced)
> > No I won't use float shit
> 
> ~then write it yourself~

Shut up its not telegram. Behave. Be civilised

Damaj301damaj-lol commented:
> > > > > @Kruna1Pate1 friend Transparent theme like Windows Terminal when ?
> > > > > ![windows-terminal-acrylic-transparent-background](https://user-images.githubusercontent.com/93857023/201646276-45417a7b-b25c-4f99-94cb-880f283a2023.jpg)
> > > > > ?
> > > > 
> > > > 
> > > > Use Termux:Float for this
> > > 
> > > 
> > > Says the one using Vanced 😂 (I also have it installed for emergency purpose but I use revanced)
> > > No I won't use float shit
> > 
> > ~then write it yourself~
> 
> Shut up its not telegram. Behave. Be civilised

Hold up what, you should be the person getting this advice

UrNightmaree commented:
It would be cool if there's an alpha option

Kruna1Pate1 commented:
> It would be cool if there's an alpha option

That what "Overlay Color" is mentioned in PR i guess.

UrNightmaree commented:
> > It would be cool if there's an alpha option
> 
> That what "Overlay Color" is mentioned in PR i guess.

When it will be push?

aicynide commented:
> I will try to do that. and i think new pr would be better for this since many changes would require.

Patiently waiting for your pr 🤗

aicynide commented:
I can see in the OP screenshot, keyboard is also transparent. How?

muttaqin1 commented:
Hello,
i am new in termux and i want to change my termux background.I am not understanding how everyone is changing their background. Can you please tell me the full process of installing termux so that i can change my termux background.

Damaj301damaj-lol commented:
> Hello,
> i am new in termux and i want to change my termux background.I am not understanding how everyone is changing their background. Can you please tell me the full process of installing termux so that i can change my termux background.

This ain't the place to learn fam, its in beta, its also unrecommended, so instead use the stable one and wait for this to get pushed

UrNightmaree commented:
> Hello, i am new in termux and i want to change my termux background.I am not understanding how everyone is changing their background. Can you please tell me the full process of installing termux so that i can change my termux background.

It's better to stick with stable version of Termux. If you still want that background support badly, you can grab it from artifacts.

anirov commented:
Unfortunately, builds with background image support are no longer available for download in [GitHub Action build](https://github.com/termux/termux-app/actions/runs/3409602990)
Please, if possible, re-upload (or share, in this way)

snooppr commented:
Termux update v0.118.1 and 0.119 f-droid, styling v 0.32.0 f-droid.
Doesn't work. There is no menu item "background" either through style or through settings. How to solve the problem?

cattokomo commented:
@snooppr As you can see, this PR is not merged to master yet. So it's obvious Termux 0.118.1 does not contain this functionality.

snooppr commented:
> @snooppr As you can see, this PR is not merged to master yet. So it's obvious Termux 0.118.1 does not contain this functionality.

and 0.119 also.

It's frustrating to see new versions being released for the sake of releasing new versions without proper documentation.

tareksander commented:
Well, in this case it's fornwall's fault that a release had to be made just for the sake of a new release.

agnostic-apollo commented:
Additionally, 0.118.1 already does provide a changelog and 0.119.0-beta.1 a link to its git history. Just gotta read. I didn't write a changelog for beta.1 because there wasn't enough time to write it for hundreds of commits, where many of changes will change in the next major beta anyways from many of my local changes.

https://github.com/termux/termux-app/releases
